Fixed #30981 -- Fixed admin changelist crash when using F() or OrderBy() expressions in admin_order_field.

hramezani opened the pull request:
[Ticket 30981](https://code.djangoproject.com/ticket/30981)

felixxm commented on django/contrib/admin/views/main.py:
```diff
@@ -308,7 +308,9 @@ def get_ordering(self, request, queryset):
                         continue  # No 'admin_order_field', skip it
-                    if hasattr(order_field, 'as_sql'):
+                    if isinstance(order_field, OrderBy):
+                        ordering.append(order_field)
```
`OrderBy` should also support reverse ordering.

felixxm replied:
We already have a  `reverse_ordering()`. Moreover this solution mutates `OrderBy` expression, so maybe:
```python
if isinstance(order_field, OrderBy):
    if pfx == '-':
        order_field = order_field.copy()
        order_field.reverse_ordering()
    ordering.append(order_field)
```

hramezani replied:
Thanks for the description. I replaced your suggested code block.

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -368,15 +368,25 @@ def test_change_list_sorting_callable_query_expression(self):
+        tests = (  # (description, index)
+            ('F("date") + DurationValue(3 days, 0:00:00)', 9),
+            ('F("date")', 12),
+            ('F("date").asc(nulls_last=True)', 13),
```
I think that property name i.e. `order_by_expression`, `order_by_f_expression`, or `order_by_orderby_expression` is a sufficient description.

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -368,15 +368,25 @@ def test_change_list_sorting_callable_query_expression(self):
+                    'Results of sorting on callable are out of order.'
+                )
 
     def test_change_list_sorting_callable_query_expression_reverse(self):
```
We should add `-12` and `-13` to this test.

hramezani commented:
@felixxm , Thanks for the review.
I changed the patch to support reverse ordering and also added a test for it.

Would it be good to add `toggle` to `django.db.models.expressions.OrderBy` :

```
    def toggle(self):
        self.descending = not self.descending
```

and use it to toggle the ordering. like: 
```
                    if isinstance(order_field, OrderBy):
                        if pfx == '-':  # Toggle ordering
                            order_field.toggle()
                        ordering.append(order_field)
```

felixxm commented:
@hramezani Thanks :+1: I pushed minor edits.
